Implicit web clip mode should choose a web clip that is visible to the application

bnham opened the pull request:
#### 77532e760a8b528addf86f4e1a73416e4dad0272
<pre>
Implicit web clip mode should choose a web clip that is visible to the application
<a href="https://bugs.webkit.org/show_bug.cgi?id=279278">https://bugs.webkit.org/show_bug.cgi?id=279278</a>
<a href="https://rdar.apple.com/135297386">rdar://135297386</a>

Reviewed by Per Arne Vollan.

The &quot;implicit web clip mode&quot; in webpushd that selects the oldest web clip associated with an origin
(added in 282857@main) doesn&apos;t totally work, because not all web clips are visible to all apps.

Instead, we now select the oldest web clip visible to the app (by checking
`trustedClientBundleIdentifiers`) and persist that selection to a plist. This also required some
other changes so that webpushd would write temp files to the appropriate directory.

Another change here is that we previously allowed multiple apps to access webpushd, but the design
from the beginning was to only allow one app to access it, so I changed methods that checked whether
apps were allowed to connect to webpushd to only allow a single app to connect.

The class that abstracts this is optimistically named WebClipCache, which doesn&apos;t actually cache a
whole lot at the moment, but will once I get around to making all the possible cache invalidation
events from system frameworks fire notifyd notifications.

I also made a drive-by fix an issue with the notification source being incorrect in
WebPushDaemon::showNotification.

* Source/WebCore/PAL/pal/spi/cocoa/CoreServicesSPI.h:
* Source/WebKit/Platform/spi/ios/UIKitSPI.h:
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.webpushd.sb.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/webpushd/PushClientConnection.h:
* Source/WebKit/webpushd/PushClientConnection.mm:
(WebPushD::PushClientConnection::subscriptionSetIdentifierForOrigin const):
(WebPushD::webClipIdentifierForOrigin): Deleted.
(WebPushD::PushClientConnection::associatedWebClipTitle const): Deleted.
* Source/WebKit/webpushd/PushService.h:
* Source/WebKit/webpushd/PushService.mm:
(WebPushD::PushService::updateSubscriptionSetState):
* Source/WebKit/webpushd/WebClipCache.h:
* Source/WebKit/webpushd/WebClipCache.mm: Added.
(WebPushD::WebClipCache::WebClipCache):
(WebPushD::webClipExists):
(WebPushD::webClipIdentifierForOrigin):
(WebPushD::WebClipCache::preferredWebClipIdentifier):
(WebPushD::loadWebClipCachePropertyList):
(WebPushD::WebClipCache::load):
(WebPushD::WebClipCache::persist):
(WebPushD::WebClipCache::visibleWebClipIdentifiers):
(WebPushD::WebClipCache::isWebClipVisible):
* Source/WebKit/webpushd/WebPushDaemon.h:
* Source/WebKit/webpushd/WebPushDaemon.mm:
(getAllowedBundleIdentifiers):
(getAllowedBundleIdentifier):
(WebPushD::WebPushDaemon::startMockPushService):
(WebPushD::WebPushDaemon::startPushService):
(WebPushD::WebPushDaemon::ensureWebClipCache):
(WebPushD::WebPushDaemon::connectionEventHandler):
(WebPushD::WebPushDaemon::updateSubscriptionSetState):
(WebPushD::WebPushDaemon::handleIncomingPush):
(WebPushD::WebPushDaemon::showNotification):
(getInstalledWebClipIdentifiers): Deleted.
(webClipExists): Deleted.
(WebPushD::platformNotificationSourceForDisplay): Deleted.
* Source/WebKit/webpushd/WebPushDaemonMain.mm:
(WebKit::WebPushDaemonMain):

Canonical link: <a href="https://commits.webkit.org/283438@main">https://commits.webkit.org/283438@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/417aff78582df27c94da398ed7e991d73c8f06e3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66266 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/45639 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/18885 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70298 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/16876 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/68384 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53438 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/17157 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/53162 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11763 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69333 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42091 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57372 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/33804 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38761 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/14753 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/15752 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/60659 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15097 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72001 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10221 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/14490 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/60485 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10253 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57436 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60787 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14614 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8441 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2068 "Passed tests") | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/41448 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/42523 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/43706 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42267 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pvollan commented on Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.webpushd.sb.in:
```diff
@@ -231,19 +231,14 @@
-
-;; Let SQLite open the parent directory of the database for fsync purposes.
-(allow file-read-data (require-all (vnode-type DIRECTORY) (home-prefix "/Library/WebKit/WebPush")))
+;; Push database and web clip cache.
+(allow file* (home-subpath "/Library/WebKit/WebPush"))
 
 ;; For MKBDeviceUnlockedSinceBoot.
 (allow iokit-open-service
     (iokit-registry-entry-class "AppleKeyStore"))
 (allow iokit-open-user-client
     (iokit-user-client-class "AppleKeyStoreUserClient"))
+
+;; Set home-relative tmpdir.
+(allow file* (home-subpath "/tmp/com.apple.webkit.webpushd"))
```
The sandbox changes look good to me.

pvollan commented on Source/WebKit/webpushd/WebClipCache.h:
```diff
@@ -0,0 +1,57 @@
+public:
+    WebClipCache(const String& path);
+
+    String preferredWebClipIdentifier(const String& bundleIdentifier, const WebCore::SecurityOriginData&);
+    HashSet<String> visibleWebClipIdentifiers(const String& bundleIdentifier);
+    bool isWebClipVisible(const String& bundleIdentifier, const String& webClipIdentifier);
```
Can these methods be const?

bnham replied:
In theory yes, although e.g. `preferredWebClipIdentifier` actually mutates some stuff behind the scenes. But I can use `mutable` to work around that.

pvollan commented on Source/WebKit/webpushd/WebClipCache.mm:
```diff
@@ -0,0 +1,253 @@
+    load();
+}
+
+static bool webClipExists(String webClipIdentifier)
```
Perhaps use const String& webClipIdentifier?

pvollan commented on Source/WebKit/webpushd/WebClipCache.mm:
```diff
@@ -0,0 +1,253 @@
+        NSError *error = nil;
+        NSData *data = [NSData dataWithContentsOfFile:path options:NSDataReadingMappedIfSafe error:&error];
+        if (!data) {
+            RELEASE_LOG_ERROR(Push, "WebClipCache::load failed to read %{public}@: %{public}@", path, error);
```
Is it ok to log these paths, or could there be privacy concerns?

bnham replied:
Should be fine, it's a system path that doesn't contain PII.

pvollan commented on Source/WebKit/webpushd/WebClipCache.mm:
```diff
@@ -0,0 +1,253 @@
+            if (!identifier)
+                continue;
+
+            if (bundleIdentifier == "com.apple.SafariViewService"_s || ![webClip respondsToSelector:@selector(trustedClientBundleIdentifiers)] || [webClip.trustedClientBundleIdentifiers containsObject:(NSString *)bundleIdentifier]) {
```
Should the identifier be added if the webClip does not respond to `trustedClientBundleIdentifiers`?

bnham replied:
Yes, it's for staging only until enough builds have that API available. Can add a comment.

pvollan reviewed (approved):
R=me.

bnham committed:
Implicit web clip mode should choose a web clip that is visible to the application
https://bugs.webkit.org/show_bug.cgi?id=279278
rdar://135297386

Reviewed by Per Arne Vollan.

The "implicit web clip mode" in webpushd that selects the oldest web clip associated with an origin
(added in 282857@main) doesn't totally work, because not all web clips are visible to all apps.

Instead, we now select the oldest web clip visible to the app (by checking
`trustedClientBundleIdentifiers`) and persist that selection to a plist. This also required some
other changes so that webpushd would write temp files to the appropriate directory.

Another change here is that we previously allowed multiple apps to access webpushd, but the design
from the beginning was to only allow one app to access it, so I changed methods that checked whether
apps were allowed to connect to webpushd to only allow a single app to connect.

The class that abstracts this is optimistically named WebClipCache, which doesn't actually cache a
whole lot at the moment, but will once I get around to making all the possible cache invalidation
events from system frameworks fire notifyd notifications.

I also made a drive-by fix an issue with the notification source being incorrect in
WebPushDaemon::showNotification.

* Source/WebCore/PAL/pal/spi/cocoa/CoreServicesSPI.h:
* Source/WebKit/Platform/spi/ios/UIKitSPI.h:
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.webpushd.sb.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/webpushd/PushClientConnection.h:
* Source/WebKit/webpushd/PushClientConnection.mm:
(WebPushD::PushClientConnection::subscriptionSetIdentifierForOrigin const):
(WebPushD::webClipIdentifierForOrigin): Deleted.
(WebPushD::PushClientConnection::associatedWebClipTitle const): Deleted.
* Source/WebKit/webpushd/PushService.h:
* Source/WebKit/webpushd/PushService.mm:
(WebPushD::PushService::updateSubscriptionSetState):
* Source/WebKit/webpushd/WebClipCache.h:
* Source/WebKit/webpushd/WebClipCache.mm: Added.
(WebPushD::WebClipCache::WebClipCache):
(WebPushD::webClipExists):
(WebPushD::webClipIdentifierForOrigin):
(WebPushD::WebClipCache::preferredWebClipIdentifier):
(WebPushD::loadWebClipCachePropertyList):
(WebPushD::WebClipCache::load):
(WebPushD::WebClipCache::persist):
(WebPushD::WebClipCache::visibleWebClipIdentifiers):
(WebPushD::WebClipCache::isWebClipVisible):
* Source/WebKit/webpushd/WebPushDaemon.h:
* Source/WebKit/webpushd/WebPushDaemon.mm:
(getAllowedBundleIdentifiers):
(getAllowedBundleIdentifier):
(WebPushD::WebPushDaemon::startMockPushService):
(WebPushD::WebPushDaemon::startPushService):
(WebPushD::WebPushDaemon::ensureWebClipCache):
(WebPushD::WebPushDaemon::connectionEventHandler):
(WebPushD::WebPushDaemon::updateSubscriptionSetState):
(WebPushD::WebPushDaemon::handleIncomingPush):
(WebPushD::WebPushDaemon::showNotification):
(getInstalledWebClipIdentifiers): Deleted.
(webClipExists): Deleted.
(WebPushD::platformNotificationSourceForDisplay): Deleted.
* Source/WebKit/webpushd/WebPushDaemonMain.mm:
(WebKit::WebPushDaemonMain):

Canonical link: https://commits.webkit.org/283438@main
